### PR TITLE
fixes #24008 - add graphql mutations

### PR DESCRIPTION
**Pull request by timogoebel**
This is part five of the graphql series.
It adds basic mutations (create, update, delete) for models.

---
Other parts:
Part 1 - JWT Auth: #5596
Part 2 - Graphql Scaffolding: #5680
Part 3 - Relay Global ID #5681
Part 4 - Connections with totalCount #5733
Part 5 - Basic mutations for model #5736

### CODE REVIEW

File: app/graphql/input_types/model_attributes.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+module InputTypes
+  class ModelAttributes < InputTypes::BaseInputObject
```
Comment (dariuszb-iRonin) [inline app/graphql/input_types/model_attributes.rb]: This is unused currently. I guess you wanted to reuse that in create and update mutations. I tried to do that by myself, however I wasn't able to find correct way of reusing it in mutations.

File: app/graphql/mutations/update_model.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
+module Mutations
+  class UpdateModel < BaseMutation
+    description 'Creates a new hardware model.'
```
Comment (dariuszb-iRonin) [inline app/graphql/mutations/update_model.rb]: Should be `Updates existing hardware model.`

File: app/graphql/mutations/create_model.rb [hunk 3]
```diff
@@ -0,0 +1,42 @@
+module Mutations
+  class CreateModel < BaseMutation
```
Comment (dariuszb-iRonin) [inline app/graphql/mutations/create_model.rb]: I suggest to put all mutations related with single resource in separate namespace(`models`). It's more readable IMO.

File: test/graphql/mutations/delete_model_mutation_test.rb [hunk 4]
```diff
@@ -0,0 +1,56 @@
+require 'test_helper'
+
+module Mutations
+  class DeleteModelMutationTest < ActiveSupport::TestCase
+    let(:model) { FactoryBot.create(:model) }
+    let(:model_id) { Foreman::GlobalId.for(model.id) }
```
Comment (dariuszb-iRonin) [inline test/graphql/mutations/delete_model_mutation_test.rb]: `Foreman::GlobalId.for` expects to receive object instead of id

File: test/graphql/mutations/update_model_mutation_test.rb [hunk 5]
```diff
@@ -0,0 +1,83 @@
+require 'test_helper'
+
+module Mutations
+  class CreateModelMutationTest < ActiveSupport::TestCase
```
Comment (dariuszb-iRonin) [inline test/graphql/mutations/update_model_mutation_test.rb]: typo, should be `UpdateModelMutationTest`

**Review by dariuszb-iRonin — COMMENTED** [review]
@timogoebel I added few comments. I also took a deep look into the code and I think we can dry it up by creating base classes for `create`, `update` and `delete` mutations. I based on your code and created PR in my fork, so you can take a look about the solution: https://github.com/dariuszb-iRonin/foreman/pull/2.

**Comment by timogoebel** [comment]
Rebased. The suggestion to dry-up the code still applies, see https://github.com/dariuszb-iRonin/foreman/pull/2/commits/7a6db41fd5025ffed2c4a994fd585298d4cb5d1a.

File: app/graphql/mutations/base_mutation.rb [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+  class BaseMutation < GraphQL::Schema::RelayClassicMutation
+    private
+
+    def authorize!(resource, action)
+      user = context[:current_user]
+      authorizer = Authorizer.new(user)
+      permission_name = resource.permission_name(action)
+
+      unless authorizer.can?(permission_name, resource)
+        raise GraphQL::ExecutionError.new(
```
Comment (xprazak2) [inline app/graphql/mutations/base_mutation.rb:37]: unauthorized

Works for now, though we should probably think about how to show errors in nice and consistent way in the future. Some errors are displayed in `data.${mutationName}.errors` while others at top level, which gives us different shape of responses. We might consider including all errors in schema as some people [suggest](https://itnext.io/the-definitive-guide-to-handling-graphql-errors-e0c58b52b5e1).

File: app/graphql/mutations/delete_mutation.rb [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+      authorize!(object, :destroy)
+
+      validate_object(object, resource_class)
+
+      User.as(context[:current_user]) do
+        object.destroy!
+
+        unless object.destroyed?
+          raise GraphQL::ExecutionError.new(
+            _('Could not destroy the record.')
```
Comment (xprazak2) [inline app/graphql/mutations/delete_mutation.rb]: Is there a way how to show nice errors telling us more about why we failed to destroy?

destroy-fail

File: app/graphql/mutations/update_mutation.rb [hunk 8]
```diff
@@ -0,0 +1,31 @@
+module Mutations
+  class UpdateMutation < BaseMutation
+    field :errors, [Types::UserError], null: false
+
+    def resolve(params)
+      object = ForemanGraphqlSchema.object_from_id(params[:id], context)
+
+      object.assign_attributes(params.except(:id))
```
Comment (xprazak2) [inline app/graphql/mutations/update_mutation.rb]: Should we `authorize!` and `validate_object` before we try to assign any attributes?

File: app/graphql/mutations/update_mutation.rb [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+      authorize!(object, :edit)
+
+      User.as(context[:current_user]) do
+        errors = if object.save
+                   []
+                 else
+                   object.errors.map do |attribute, message|
```
Comment (xprazak2) [inline app/graphql/mutations/update_mutation.rb]: Maybe extract to stay DRY as both `UpdateMutation` and `CreateMutation` use same error handling.

File: app/graphql/types/user_error.rb [hunk 10]
```diff
@@ -0,0 +1,11 @@
+module Types
+  class UserError < BaseObject
```
Comment (xprazak2) [inline app/graphql/types/user_error.rb]: I find the class name a bit misleading, only after reading the description it was clear that the error is not related to `User` model. How about ValidationError, AttributeError or something similar?

File: app/graphql/types/user_error.rb [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+    description 'A user-readable error'
+
+    field :message, String, null: false,
+      description: 'A description of the error'
+
+    field :path, [String], null: true,
```
Comment (xprazak2) [inline app/graphql/types/user_error.rb]: Not sure if having just attribute name would be enough. On the other hand, this could be useful for nested attributes if we need them, so I think we can keep it as it is for now.

**Comment by xprazak2** [comment]
Nice! Based on my initial testing, everything seems to work. I left a couple of comments, mostly cosmetic stuff.

File: app/graphql/mutations/update_mutation.rb (hunk 9 above)
Comment (timogoebel) [reply app/graphql/mutations/update_mutation.rb]: Extracted to `save_object` method.

**Comment by timogoebel** [comment]
@xprazak2: Thanks for your comments. I continued to work on this PR.

- `required?` is now automatically detected
- `resource_class` is now automatically detected
- a `DeleteMutation` now shows proper error messages
- and I addressed your comment

File: app/graphql/mutations/delete_mutation.rb [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+
+    def resolve(id:)
+      object = load_object_by(id: id)
+      authorize!(object, :destroy)
+
+      User.as(context[:current_user]) do
+        errors = if object.destroy
+                   []
+                 else
+                   map_errors_to_path(resource)
```
Comment (xprazak2) [inline app/graphql/mutations/delete_mutation.rb]: I think this should be `object` as `resource` is not defined. But when I change it, I get:

```
2019-04-01T08:15:45 [I|aud|a4efced6] Model (4) destroy event on name Blueberry
2019-04-01T08:15:45 [I|aud|a4efced6] Model (4) destroy event on info berry hw model updated
2019-04-01T08:15:45 [I|aud|a4efced6] Model (4) destroy event on vendor_class Awesome vendor
2019-04-01T08:15:45 [I|aud|a4efced6] Model (4) destroy event on hardware_model 4
2019-04-01T08:15:45 [E|app|a4efced6] You may not destroy Blueberry as it is in use!
2019-04-01T08:15:45 [I|app|a4efced6] Current user set to admin (admin)
2019-04-01T08:15:45 [I|app|a4efced6] Completed 500 Internal Server Error in 49ms (ActiveRecord: 8.7ms)
 | 
 | 
2019-04-01T08:15:45 [F|app|a4efced6]   
2019-04-01T08:15:45 [F|app|a4efced6] NoMethodError (undefined method `object' for #<Hash:0x007fbc448f25e8>
 | Did you mean?  object_id):
2019-04-01T08:15:45 [F|app|a4efced6]   
2019-04-01T08:15:45 [F|app|a4efced6] app/controllers/api/graphql_controller.rb:47:in `execute_single_graphql_query'
 | app/controllers/api/graphql_controller.rb:18:in `execute'
 | app/controllers/concerns/foreman/controller/timezone.rb:10:in `set_timezone'
 | app/models/concerns/foreman/thread_session.rb:32:in `clear_thread'
 | lib/foreman/middleware/telemetry.rb:10:in `call'
 | lib/foreman/middleware/catch_json_parse_errors.rb:9:in `call'
 | lib/foreman/middleware/logging_context_session.rb:22:in `call'
 | lib/foreman/middleware/logging_context_request.rb:11:in `call'
 | /home/vagrant/projects/katello/lib/katello/prevent_json_parsing.rb:12:in `call'

```

**Comment by xprazak2** [comment]
Awesome work, just one problem when handling errors on delete.

**Comment by timogoebel** [comment]
> Awesome work, just one problem when handling errors on delete.

@xprazak2: Thanks, updated the code and added `graphql_name` to the mutations as it's not properly set automatically.

**Review by xprazak2 — DISMISSED** [review]
Works smoothly, test failure unrelated

**Comment by timogoebel** [comment]
@xprazak2: Thanks for reviewing and testing! If we can get this into 1.22, we'd have a nice basis that plugins can work with. :-)

**Comment by xprazak2** [comment]
Sure, could you rebase pls?

**Comment by timogoebel** [comment]
@xprazak2: Rebased. I used the chance to dry up the required attribute detection by introducing a small service class to do that. Also added a bit of test coverage for this.

**Comment by timogoebel** [comment]
Test failures are unrelated.

**Review by xprazak2 — APPROVED** [review]

**Comment by xprazak2** [comment]
Thanks!
